File: app/src/main/java/co/smartreceipts/android/model/impl/ImmutablePaymentMethodImpl.java
package co.smartreceipts.android.model.impl;

import android.content.res.Resources;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.google.common.base.Preconditions;

import co.smartreceipts.android.model.PaymentMethod;
import co.smartreceipts.android.model.factory.PaymentMethodBuilderFactory;
import co.smartreceipts.android.sync.model.SyncState;
import co.smartreceipts.android.sync.model.impl.DefaultSyncState;

/**
 * An immutable implementation of {@link co.smartreceipts.android.model.PaymentMethod}.
 *
 * @author Will Baumann
 */
public final class ImmutablePaymentMethodImpl implements PaymentMethod {

    // TODO: 15.07.2017 maybe would be good to delete 'Unspecified' payment method
    // TODO: 16.07.2017 check tests
    public static final PaymentMethod NONE = new PaymentMethodBuilderFactory().setMethod(Resources.getSystem().getString(android.R.string.untitled)).build();

    private final int mId;
    private final String mMethod;
    private final SyncState mSyncState;

    public ImmutablePaymentMethodImpl(int id, @NonNull String method) {
        this(id, method, new DefaultSyncState());
    }

    public ImmutablePaymentMethodImpl(int id, @NonNull String method, @NonNull SyncState syncState) {
        mId = id;
        mMethod = Preconditions.checkNotNull(method);
        mSyncState = Preconditions.checkNotNull(syncState);
    }

    private ImmutablePaymentMethodImpl(final Parcel in) {
        mId = in.readInt();
        mMethod = in.readString();
        mSyncState = in.readParcelable(getClass().getClassLoader());
    }

    /**
     * @return - the database primary key id for this method
     */
    @Override
    public int getId() {
        return mId;
    }

    @Override
    @NonNull
    public String getMethod() {
        return mMethod;
    }

    @NonNull
    @Override
    public SyncState getSyncState() {
        return mSyncState;
    }

    @Override
    public String toString() {
        return mMethod;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ImmutablePaymentMethodImpl)) return false;

        ImmutablePaymentMethodImpl that = (ImmutablePaymentMethodImpl) o;

        if (mId != that.mId) return false;
        return (mMethod.equals(that.mMethod));
    }

    @Override
    public int hashCode() {
        int result = mId;
        result = 31 * result + mMethod.hashCode();
        return result;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel out, final int flags) {
        out.writeInt(mId);
        out.writeString(mMethod);
        out.writeParcelable(mSyncState, flags);
    }

    public static Creator<ImmutablePaymentMethodImpl> CREATOR = new Creator<ImmutablePaymentMethodImpl>() {

        @Override
        public ImmutablePaymentMethodImpl createFromParcel(Parcel source) {
            return new ImmutablePaymentMethodImpl(source);
        }

        @Override
        public ImmutablePaymentMethodImpl[] newArray(int size) {
            return new ImmutablePaymentMethodImpl[size];
        }

    };

}

File: app/src/main/kotlin/co/smartreceipts/android/model/impl/ImmutablePaymentMethodImpl.kt
package co.smartreceipts.android.model.impl

import android.content.res.Resources
import co.smartreceipts.android.model.PaymentMethod
import co.smartreceipts.android.model.factory.PaymentMethodBuilderFactory
import co.smartreceipts.android.sync.model.SyncState
import co.smartreceipts.android.sync.model.impl.DefaultSyncState
import kotlinx.android.parcel.Parcelize
import java.util.*

/**
 * An immutable implementation of [PaymentMethod].
 *
 * @author Will Baumann
 */
@Parcelize
class ImmutablePaymentMethodImpl @JvmOverloads constructor (
    override val id: Int,
    override val uuid: UUID,
    override val method: String,
    override val syncState: SyncState = DefaultSyncState(),
    override val customOrderId: Long = 0

) : PaymentMethod {

    override fun toString() = method

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is ImmutablePaymentMethodImpl) return false

        val that = o as ImmutablePaymentMethodImpl?

        if (id != that!!.id) return false
        if (uuid != that.uuid) return false
        return if (customOrderId != that.customOrderId) false else method == that.method
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + uuid.hashCode()
        result = 31 * result + method.hashCode()
        result = 31 * result + (customOrderId xor customOrderId.ushr(32)).toInt()
        return result
    }

    override fun compareTo(paymentMethod: PaymentMethod): Int {
        return customOrderId.compareTo(paymentMethod.customOrderId)
    }

    companion object {
        val NONE = PaymentMethodBuilderFactory().setMethod(Resources.getSystem().getString(android.R.string.untitled)).build()
    }
}
